controllers/monitoringconfig_controller.go

```diff
@@ -305,3 +318,78 @@ func deduplicateRemoteWrites(remoteWrites []interface{}) []interface{} {
 	})
 	return remoteWrites
 }
+
+// cleanupSecrets deletes all secrets in the monitoring namespace that belong to the given MonitoringConfig.
+func (r *MonitoringConfigReconciler) cleanupSecrets(ctx context.Context, mc *everestv1alpha1.MonitoringConfig) error {
+	// List secrets in the monitoring namespace that belong to this MonitoringConfig.
+	secrets := &corev1.SecretList{}
+	err := r.List(ctx, secrets, &client.ListOptions{
+		Namespace: r.monitoringNamespace,
+		LabelSelector: labels.SelectorFromSet(map[string]string{
+			monitoringConfigRefNameLabel:      mc.GetName(),
+			monitoringConfigRefNamespaceLabel: mc.GetNamespace(),
+		}),
+	})
+	if err != nil {
+		return err
+	}
+	for _, secret := range secrets.Items {
+		if err := r.Delete(ctx, &secret); err != nil {
+			return err
+		}
+	}
+	// Remove the finalizer from the MonitoringConfig.
+	if controllerutil.RemoveFinalizer(mc, monitoringConfigSecretCleanupFinalizer) {
+		return r.Update(ctx, mc)
+	}
+	return nil
+}
+
+// reconcileSecret copies the source MonitoringConfig secret onto the monitoring namespace.
+// Returns the name of the newly created/updated secret.
+func (r *MonitoringConfigReconciler) reconcileSecret(ctx context.Context, mc *everestv1alpha1.MonitoringConfig) (string, error) {
+	// If the MonitoringConfig is in the monitoring namespace, we don't have to do any additional work.
+	// Use the secret as-is.
+	if mc.GetNamespace() == r.monitoringNamespace {
+		return mc.Spec.CredentialsSecretName, nil
+	}
+
+	// Get the secret in the MonitoringConfig namespace.
+	src := &corev1.Secret{}
+	err := r.Get(ctx, types.NamespacedName{
+		Name:      mc.Spec.CredentialsSecretName,
+		Namespace: mc.GetNamespace(),
+	}, src)
+	if err != nil {
+		return "", err
+	}
+
+	// Create a copy in the monitoring namespace.
+	dstSecretName := src.GetName() + "-" + src.GetNamespace()
+	dst := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      dstSecretName,
+			Namespace: r.monitoringNamespace,
+		},
+	}
+	if _, err := controllerutil.CreateOrUpdate(ctx, r.Client, dst, func() error {
+		dst.Data = src.DeepCopy().Data
+		// Attach labels to identify the parent MonitoringConfig.
+		// This is useful for garbage collection, since we cannot have cross-namespace OwnerRefs.
+		labels := dst.GetLabels()
+		if labels == nil {
+			labels = make(map[string]string)
+		}
+		labels[monitoringConfigRefNameLabel] = mc.GetName()
+		labels[monitoringConfigRefNamespaceLabel] = mc.GetNamespace()
+		dst.SetLabels(labels)
+		return nil
+	}); err != nil {
+		return "", err
+	}
+	// Add a clean-up finalizer in the parent MonitoringConfig.
+	if controllerutil.AddFinalizer(mc, monitoringConfigSecretCleanupFinalizer) {
+		return dstSecretName, r.Update(ctx, mc)
+	}
```

Comment:
should this method check `mc.GetDeletionTimestamp().IsZero()` before adding the finalizer? I mean, is there a chance it will be added again here after it was deleted?

Reply:
I don't think so..

[Here](https://github.com/percona/everest-operator/pull/500/files#diff-1d6b7319d3daf756d50f49cb113d91450fbe3022146f58e9793a931a8f4e01a0R84) we're already checking for the deletion timestamp at the top level `Reconcile`. If it is not zero, we cleanup and return early, even before `reconcileSecrets` gets called. So I don't think the finalizer will be added again